### DISCUSSION TITLE
Proposal for including the ParameterPath property in the EventToCommandBehavior class to retrieve the CommandParameter value from a property mapped within the EventArgs object

### DISCUSSION THREAD

**Discussion by egvijayanand** [CommunityToolkit/Maui#2365] [New Feature Discussions]
Proposal for including the ParameterPath property in the EventToCommandBehavior class to retrieve the CommandParameter value from a property mapped within the EventArgs object.

This EventToCommandBehavior is undoubtedly one of the most useful types in the MVVM context as it is the de facto option when there is no support for commands.

Often, the parameter value required for the ViewModel command is encapsulated within the EventArg object.

A converter is necessary to complete the task. However, most converters perform a similar function: they assign (map) the value for the parameter from a property.

Agree that a generic version of the EventToCommandBehavior exists, yet it is not fully functional in scenarios where ViewModels are defined within a library and shared across different AppModels. For example, using .NET MAUI for Mobile Apps and WPF for Desktop Apps, but VM is shared. The EventArgs types are UI-specific.

To address such scenarios, it is proposed to add the ParameterPath property, which is simply the property of interest.

Kindly let me know your thoughts on this.

**Change:**

Below is the code that I wrote for this (only relevant section is shown):

```cs
public class EventToCommandBehavior : BaseBehavior<VisualElement>
{
    public static readonly BindableProperty ParameterPathProperty
        = BindableProperty.Create(nameof(ParameterPath), typeof(string), typeof(EventToCommandBehavior), default);

    public string ParameterPath
    {
        get => (string)GetValue(ParameterPathProperty);
        set => SetValue(ParameterPathProperty, value);
    }

    protected virtual void OnTriggerHandled(object? sender = null, object? eventArgs = null)
    {
        // The priority is maintained.
        // Explicit Parameter value, invoke the Converter, process the Path, eventArgs itself (default)
        var parameter = CommandParameter
                ?? EventArgsConverter?.Convert(eventArgs, typeof(object), null, CultureInfo.InvariantCulture)
                ?? (string.IsNullOrWhiteSpace(ParameterPath) ? eventArgs : GetParameterValue(eventArgs));

        var command = Command;

        if (command?.CanExecute(parameter) ?? false)
        {
            command.Execute(parameter);
        }

        object? GetParameterValue(object? eventArgs)
        {
            object? propertyValue = eventArgs;
            PropertyInfo? propertyInfo;

            // For nested properties
            foreach (var propertyPathPart in ParameterPath.Split('.'))
            {
                propertyInfo = propertyValue?.GetType().GetProperty(propertyPathPart);
                propertyValue = propertyInfo?.GetValue(propertyValue);
            }

            return propertyValue;
        }
    }
}
```

**Usage:**

```xml
<HybridWebView
    x:Name="hwv">
    <HybridWebView.Behaviors>
        <mct:EventToCommandBehavior
            Command="{Binding ShowMessageCommand}"
            EventName="RawMessageReceived"
            ParameterPath="Message" />
    </HybridWebView.Behaviors>
</HybridWebView>
```
```cs
var hwv = new HybridWebView();
hwv.Behaviors.Add(new EventToCommandBehavior()
{
    EventName = nameof(HybridWebView.RawMessageReceived),
    ParameterPath = nameof(HybridWebViewRawMessageReceivedEventArgs.Message),
});
```

**Comment by egvijayanand** [comment]
@jfversluis @pictos Please share your thoughts on this.

**Reply by bijington** [reply]
Will this play nicely with trimming/AOT support given there is reflection being used? If not perhaps it could be modified to allow the developer to provide a `Parameter` property that could be controlled through data binding?

**Reply by egvijayanand** [reply]
`ParameterPath` is a bindable property. Developers can data bind it.

But the event info itself is retrieved using Reflection only.

https://github.com/CommunityToolkit/Maui/blob/9741f1ad4cd24ea00eb99e0bc3f09c47a6afb73e/src/CommunityToolkit.Maui/Behaviors/EventToCommandBehavior.shared.cs#L106
